Store creation M2 - domain selection: empty/error/loading states, design updates

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Parts of #8045 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

After the previous basic UI integration, this PR updated the design and handled different states of the domain suggestions.

Design updates include:
- Updated the large title from a label in the view to the large title style for the navigation bar, so that it doesn't take up much vertical space on landscape and shrinks when scrolling up
- Landscape layout:
  - Previously, `List` was used for the domain results but it had its own scrolling behavior and it was hard to fit the height to its content. This PR changed it to show the domain suggestions in a `LazyVStack` (there are always max 20 results without infinite scroll) contained in a `ScrollView`. This way, the user can always scroll to the bottom of the domain list
- Search text field (`SearchHeader` component): it's currently used in `ProductSelector`and `FilterListSelector`, I tested these two use cases to make sure there are no visual changes. I added a `Customizations` struct that can be passed to the initializer, with the styles in current use cases as default
- `DomainRowView`: added padding to the row

States of domain suggestions:
- **Empty query**: whenever the query is empty (including empty spaces), a placeholder image is shown - please note that the image asset needs some updates to support dark mode HyVloP5FipZzyPVenH2euI-fi-2453%3A99694 as a separate subtask
- **Error**: an error message is shown. If it's a `DotcomError.unknown` error, the `message` string is shown (should be localized)
- **Loading**: a `ProgressView` spinner is shown
- **Results**: the domain suggestions are shown

I updated the SwiftUI previews to include previews for these 4 states for faster design updates during development. Please feel free to suggest anything, I haven't worked on SwiftUI extensively for a while.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

- Log in if needed
- Go to the Menu tab, and tap `Switch store`
- On the store picker, tap `+ Add a store`
- Tap `Create a new store` --> the domain selector should be shown with a placeholder image
- Enter some valid query for domain suggestions --> a spinner should be shown while loading. after a bit, some results should be shown
- Select a domain --> a checkmark should be shown on the row with bold text, and `Continue` CTA should be shown at the bottom
- Enter some invalid query (e.g. `!!!!`) --> the `Continue` CTA should be hidden, and a spinner should be shown while loading. after a bit, an error message should be shown like `Domain searches only support the following characters...`
- Feel free to test the above steps in landscape mode

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->

Please feel free to suggest any other improvements.

empty query | loading | error | results
-- | -- | -- | --
![Simulator Screen Shot - iPhone 14 - 2022-11-11 at 10 00 10](https://user-images.githubusercontent.com/1945542/201246003-ae271f13-680e-4ba4-800e-411cdde29f23.png) | ![Simulator Screen Shot - iPhone 14 - 2022-11-11 at 10 00 23](https://user-images.githubusercontent.com/1945542/201246010-f91ecc95-ed49-4ca5-b58c-7604fe948b22.png) | ![Simulator Screen Shot - iPhone 14 - 2022-11-11 at 10 00 28](https://user-images.githubusercontent.com/1945542/201246014-cbccfe55-d52d-489d-942c-6ac1e3747404.png) | ![Simulator Screen Shot - iPhone 14 - 2022-11-11 at 10 01 05](https://user-images.githubusercontent.com/1945542/201246016-20fcfd97-4c4d-43cc-966a-1324448d7793.png)


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

jaclync committed:
Use navigation bar large title for store creation domain selector.

jaclync committed:
Handle error by showing an error message. Improve landscape layout by using a `LazyVStack` instead of `List`. Improve loading state.

jaclync committed:
Add domain search placeholder image asset.

jaclync committed:
Show placeholder image when search query is empty.

jaclync committed:
Add customizations to `SearchHeader` component for `DomainSelectorView` to match design.

jaclync committed:
Comment and access control updates.

jaclync committed:
Update unit tests to wait for reactive property updates.

jaclync committed:
Update domain search placeholder asset - still might need a dark mode version.

annchichi commented:
Great work @jaclync! That's super fast. 

- It'd be odd when the graph feels half-cut. I wonder if we can have the keyboard up with the blinking typing indicator in the search bar on when the query is empty?
- There were some spacing issues, so I noted this with this screenshot. 
- The suggestion can be smaller with SF pro text 13 regular. 

<img width="1033" alt="CleanShot 2022-11-10 at 19 00 32@2x" src="https://user-images.githubusercontent.com/13378873/201253532-4c3ed4d7-f626-4b0c-b43d-2e3c4fdf98bb.png">

Thank you!

jaclync commented:
Thanks @annchichi for the feedback! I'm leaving these improvements for a separate PR later, since this PR is already in review. I added them as subtasks in the issue description https://github.com/woocommerce/woocommerce-ios/issues/8045.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/Domains/DomainSelectorView.swift:
```diff
@@ -70,49 +73,98 @@ struct DomainSelectorView: View {
+                                }
+                            }
+                        }
+                    }
```
We can simply decide visibility of the placeholder image based on `searchTerm`. That way we won't need to update the image in the view model. We can also avoid the else-if blocks by using `renderredIf` - the results `LazyVStack` will not be rendered if the domain list is empty so there's no need to put it in the `else` block I believe.
```Swift
// Placeholder image when search query is empty.
HStack {
    Spacer()
    Image(uiImage: UIImage.domainSearchPlaceholderImage)
    Spacer()
}
.renderedIf(viewModel.searchTerm.isEmpty)

HStack {
    Spacer()
    ProgressView()
    Spacer()
}
.renderedIf(viewModel.isLoadingDomainSuggestions)

if let errorMessage = viewModel.errorMessage {
    // Error message when there is an error loading domain suggestions.
    Text(errorMessage)
        .padding(Layout.defaultPadding)
}

LazyVStack {
    ForEach(viewModel.domains, id: \.self) { domain in
        Button {
            selectedDomainName = domain
        } label: {
            VStack(alignment: .leading) {
                DomainRowView(viewModel: .init(domainName: domain,
                                                searchQuery: viewModel.searchTerm,
                                                isSelected: domain == selectedDomainName))
                Divider()
                    .frame(height: Layout.dividerHeight)
                    .padding(.leading, Layout.defaultHorizontalPadding)
            }
        }
    }
}
```

jaclync replied:
Nice suggestions, I forgot about `renderedIf`! The if/else style ensures that only one conditional UI is shown at a time. I tested the suggested changes, and I'll have to change the timing when errors are reset and maybe more to make sure only one UI is shown at the same time. I think I'll keep the if/else style for now, but I replaced the `viewModel.placeholderImage` with `viewModel.searchTerm.isEmpty` in https://github.com/woocommerce/woocommerce-ios/pull/8092/commits/42d811f76e90c7f1ff727a72e5fdc6bc8dc0acbf thanks to your suggestion!

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/Domains/DomainSelectorViewModel.swift:
```diff
@@ -11,30 +12,55 @@ final class DomainSelectorViewModel: ObservableObject {
+    @Published private(set) var isLoadingDomainSuggestions: Bool = false
+
+    /// Placeholder image is set when the search term is empty. Otherwise, the placeholder image is `nil`.
+    @Published private(set) var placeholderImage: UIImage?
```
As comment above, we should not need this to decide the visibility of the placeholder image - the `searchTerm` should be sufficient.

jaclync replied:
Removed in https://github.com/woocommerce/woocommerce-ios/pull/8092/commits/42d811f76e90c7f1ff727a72e5fdc6bc8dc0acbf 👍

itsmeichigo reviewed (approved):
Overall this works great, and the code looks good to me. I have a few non-blocking comments:

<img src="https://user-images.githubusercontent.com/5533851/201269711-7ccd5411-eb1e-4933-a3d7-c9bc7d3df9b2.png" width=320 />
Here it seems to me that the padding around the search icon is a bit large, while there seems to be missing some padding on the right of the text field. This may look out of place for devices with small screen sizes.


Also, we probably want to dismiss the keyboard after selecting any item and when the list is being dragged. That will help avoid situations when the list is not scrollable due to the keyboard taking up all of the space:

![Simulator Screen Shot - iPhone 13 Pro - 2022-11-11 at 12 25 00](https://user-images.githubusercontent.com/5533851/201269856-b11a28ce-7c9d-43fd-9a1b-1cfe3247c25d.png)

jaclync committed:
Merge branch 'trunk' into feat/8045-domain-selector-updates

* trunk: (41 commits)
  Freeze strings for localization
  Update WordPressAuthenticator pod to stable 4.2.0 version
  Update metadata strings
  Update release notes for 11.2
  Release Notes: add new section for next version (11.3)
  Update draft release notes for 11.2.
  Bump version number
  Bump version number
  Update metadata translations
  Update app translations – `Localizable.strings`
  Updated release notes for 11.1 to include product preview
  Add Site dependency with valid frame-nonce to tests
  Update release notes
  Update tests
  Remove productsOnboarding feature flag
  Update release notes for 11.2
  Change status_code type from String to Int64
  Address warning about `try` used on a method that does not `throws`
  Send product_detail_preview_failed event from WebKitViewController
  Add product_detail_preview_failed event
  ...

jaclync committed:
Replace `DomainSelectorViewModel.placeholderImage` with `searchTerm` check to show a placeholder image.

jaclync commented:
Thanks for the review & testing feedback @itsmeichigo! I addressed your comments, and added your design suggestions to https://github.com/woocommerce/woocommerce-ios/issues/8045 as future subtasks. I'll come back to the design enhancements after the main store creation flow is done. Merging the PR now.
